### DISCUSSION TITLE
Defining colors: Displacy entity recognizer

### DISCUSSION THREAD

**Discussion by balachander1964** [explosion/spaCy#11651] [Help: Coding & Implementations] [answered]
Hi, Could you please let me know how to define the background and foreground colors when using displacy entity recognizer? I am more interested in using solid colors. Thank you for your reply in advance. Bala

**Comment by ljvmiranda921** [comment, marked as answer]
Hi @balachander1964 , you can override the color [via the displaCy API](https://spacy.io/usage/visualizers#ent). You need to pass a dictionary (where each key is an entity-type with its corresponding color value) in the `colors` parameter. 

However, there's no option to do this for the background color. You might need to use HTML/CSS templates as a workaround. The displaCy templates [are made of HTML tags](https://github.com/explosion/spaCy/blob/master/spacy/displacy/templates.py). You can copy what's in `TPL_ENTS` and `TPL_ENT`, then create a `div` that sets the `background-color` to something you like, Afterwards, you can pass it [in the `template` option of the visualizer](https://spacy.io/api/top-level#displacy_options-ent).

**Reply by balachander1964** [reply]
Hi,Thank you for your reply. Could you please forward some sample code so that I can use this as a base. I appreciate your kind help in advance. Bala

**Reply by balachander1964** [reply]
Hi, I tried defining my own options when calling displacy.render. I figured out that displacy ignores the options altogether and defines all ents using pre-defined colurs. Also, adding my templates crashes and raises an exception. I request you to let me know how to fix this. I am forwrding the source code and the generated html. Thanks in advanced. Bala.

<<Python Source code here:>>
#imports
import os, sys
import pandas as PD
import numpy as NP
import spacy
from spacy import matcher
from spacy.matcher import matcher, dependencymatcher, phrasematcher
from spacy.matcher.matcher import Matcher
from spacy.matcher.phrasematcher import PhraseMatcher
from spacy.matcher.dependencymatcher import DependencyMatcher
from spacy import displacy

import logging

file_name = (__file__).replace('.py', '') + '.txt'
fmt = '[%(asctime)s-%(msecs)d]: {%(exec_file)s: %(funcName)s: %(lineno)d} - %(levelname)s - %(message)s'

logging.basicConfig(
  level = logging.INFO,
  filename = file_name ,
  filemode = 'w',
  format= '[%(asctime)s-%(msecs)d]: {%(filename)s: %(funcName)s: %(lineno)d} - %(levelname)s - %(message)s',
  datefmt='%Y-%m-%d %H:%M:%S'
  )

_entry_stmt = 'Entering the method.'
_leaving_stmt ='Leaving the method.'
_text = '''He bought a new Toyota Corola car last year'''
_model = 'en_core_web_lg'

_template = """
<div class="entities" style="line-height: 2.5; direction: {dir}">{content}</div>
<mark class="entity" style="background: {bg}; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;">
  {text}
  <span style="font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem">{label}{kb_link}</span>
</mark>
"""

_ents_list = ["GPE", "PERSON", "ORG", "PRODUCT", "DATE", "TIME", "QUANTITY", "ORDINAL", "CAARDINAL"]
_colors_dict = {
  'ORG' : '#800000',
  'GPE' : '#800000',
  'PRODUCT' : '#000080',
  'PERSON' : '#000080',
  'DATE' : '#006400',
  'TIME' : 'Dark #006400',
  'CARDINAL' : '#9370DB',
  'ORDINAL' : '#9370DB',
  }
def main():
  '''Main entry point'''
  logging.debug( _entry_stmt)

  nlp = spacy.load(_model)
  doc = nlp(_text)
  try:
#    [print(ent.text, ent.label_) for ent in doc.ents]
    html = displacy.render([doc], style = 'ent', options = _options ,  jupyter = False)

    with open( './myfile.html', mode = 'w') as my_file:
      my_file.write(html)
      my_file.flush()
  except:
    logging.exception('An exception in the code.')

  logging.debug(_leaving_stmt)
  return

'''
  Required for all python programs.
'''
if __name__ == '__main__':
  logging.info('Starting the application')
  print('Starting the application.')
  main()
  print('Leaving the Application.\nPress <Enter> to exit.')
  logging.info('leaving the application.')
  input()
<<End of source code>>

<<Generated html here>>
<div class="entities" style="line-height: 2.5; direction: ltr">He bought a new 
<mark class="entity" style="background: yellow; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;">
    Toyota
    <span style="font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem">ORG</span>
</mark>
 
<mark class="entity" style="background: #ADFF2F; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;">
    Corola
    <span style="font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem">PRODUCT</span>
</mark>
 car 
<mark class="entity" style="background: #ADFF2F; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;">
    last year
    <span style="font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem">DATE</span>
</mark>
</div>
<<End of generated html>>

**Comment by balachander1964** [comment]
Hi, Got your reply working. I appreciate your quick reply. Thank you very much. Bala

**Reply by balachander1964** [reply]
Hi, I am reforwarding the source code:
<<Python Source code here:>>
#imports
import os, sys
import pandas as PD
import numpy as NP
import spacy
from spacy import matcher
from spacy.matcher import matcher, dependencymatcher, phrasematcher
from spacy.matcher.matcher import Matcher
from spacy.matcher.phrasematcher import PhraseMatcher
from spacy.matcher.dependencymatcher import DependencyMatcher
from spacy import displacy

import logging

file_name = (__file__).replace('.py', '') + '.txt'
fmt = '[%(asctime)s-%(msecs)d]: {%(exec_file)s: %(funcName)s: %(lineno)d} - %(levelname)s - %(message)s'

logging.basicConfig(
  level = logging.INFO,
  filename = file_name ,
  filemode = 'w',
  format= '[%(asctime)s-%(msecs)d]: {%(filename)s: %(funcName)s: %(lineno)d} - %(levelname)s - %(message)s',
  datefmt='%Y-%m-%d %H:%M:%S'
  )

_entry_stmt = 'Entering the method.'
_leaving_stmt ='Leaving the method.'
_text = '''He bought a new Toyota Corola car last year'''
_model = 'en_core_web_lg'

_template = """
<div class="entities" style="line-height: 2.5; direction: {dir}">{content}</div>
<mark class="entity" style="background: {bg}; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;">
  {text}
  <span style="font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem">{label}{kb_link}</span>
</mark>
"""

_ents_list = ["GPE", "PERSON", "ORG", "PRODUCT", "DATE", "TIME", "QUANTITY", "ORDINAL", "CAARDINAL"]
_colors_dict = {
  'ORG' : '#800000',
  'GPE' : '#800000',
  'PRODUCT' : '#000080',
  'PERSON' : '#000080',
  'DATE' : '#006400',
  'TIME' : '#006400',
  'CARDINAL' : '#9370DB',
  'ORDINAL' : '#9370DB',
  }
def main():
  '''Main entry point'''
  logging.debug( _entry_stmt)

  nlp = spacy.load(_model)
  doc = nlp(_text)
  try:
#    [print(ent.text, ent.label_) for ent in doc.ents]
    html = displacy.render([doc], style = 'ent', options = _options ,  jupyter = False)

    with open( './myfile.html', mode = 'w') as my_file:
      my_file.write(html)
      my_file.flush()
  except:
    logging.exception('An exception in the code.')

  logging.debug(_leaving_stmt)
  return

'''
  Required for all python programs.
'''
if __name__ == '__main__':
  logging.info('Starting the application')
  print('Starting the application.')
  main()
  print('Leaving the Application.\nPress <Enter> to exit.')
  logging.info('leaving the application.')
  input()
<<End of source code>>

<<Generated html here>>
<div class="entities" style="line-height: 2.5; direction: ltr">He bought a new 
<mark class="entity" style="background: yellow; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;">
    Toyota
    <span style="font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem">ORG</span>
</mark>
 
<mark class="entity" style="background: #ADFF2F; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;">
    Corola
    <span style="font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem">PRODUCT</span>
</mark>
 car 
<mark class="entity" style="background: #ADFF2F; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;">
    last year
    <span style="font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem">DATE</span>
</mark>
</div>
<<End of generated html>>

**Reply by ljvmiranda921** [reply]
Hi @balachander1964 , can you format your code first so that it's easier for us to help you debug the problem? You can enclose your code with triple backticks  (```) to define codeblocks. Ideally, we just need a minimal reproducible example. You don't have to include the whole source, just the necessary parts related to your displaCy question.

**Reply by balachander1964** [reply]
Hi ljvMirandan, I could not reply immediately since I was out of town. I am forwarding the code block now. Awaiting your reply. Thank you. Bala

![image](https://user-images.githubusercontent.com/68279037/197751143-30599984-f388-4410-bec8-dc315f87a8e1.png)

**Reply by balachander1964** [reply]
![image](https://user-images.githubusercontent.com/68279037/197751365-86b57d84-0959-4ddb-91fb-44c50bae7bc5.png)

**Reply by ljvmiranda921** [reply]
Hi @balachander1964 , I think you're meant to edit the original message and update the codeblocks (or source code) there. You don't need to send a screenshot. Again, you can use triple backticks (```) to enclose code properly. We also appreciate if you provide us a minimum reproducible example, without the extra logging information so we can help debug your problem.

**Comment by balachander1964** [comment]
```
import spacy
from spacy import displacy

_text = '''He bought a new Toyota Corola car last year'''
_model = 'en_core_web_lg'

_template = """
<div class="entities" style="line-height: 2.5; direction: {dir}">{content}</div>
<mark class="entity" style="background: {bg}; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;">
{text}
<span style="font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem">{label}{kb_link}</span>
</mark>
"""

_ents_list = ["GPE", "PERSON", "ORG", "PRODUCT", "DATE", "TIME", "QUANTITY", "ORDINAL", "CAARDINAL"]
_colors_dict = {
'ORG' : '#800000',
'GPE' : '#800000',
'PRODUCT' : '#000080',
'PERSON' : '#000080',
'DATE' : '#006400',
'TIME' : 'Dark #006400',
'CARDINAL' : '#9370DB',
'ORDINAL' : '#9370DB',
}
nlp = spacy.load(_model)
doc = nlp(_text)
html = displacy.render([doc], style = 'ent', options = _options ,jupyter = False)
with open( './myfile.html', mode = 'w') as my_file:
  my_file.write(html)
  my_file.flush()
```

**Comment by balachander1964** [comment]
Please ignore my previous text which I forwarded just few minutes ago. Here is the code enclosed within the markers.

```
import spacy
from spacy import displacy

_text = '''He bought a new Toyota Corola car last year'''
_model = 'en_core_web_lg'

_template = """
<div class="entities" style="line-height: 2.5; direction: {dir}">{content}</div>
<mark class="entity" style="background: {bg}; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;">
{text}
<span style="font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem">{label}{kb_link}</span>
</mark>
"""

_ents_list = ["GPE", "PERSON", "ORG", "PRODUCT", "DATE", "TIME", "QUANTITY", "ORDINAL", "CAARDINAL"]
_colors_dict = {
'ORG' : '#800000',
'GPE' : '#800000',
'PRODUCT' : '#000080',
'PERSON' : '#000080',
'DATE' : '#006400',
'TIME' : 'Dark #006400',
'CARDINAL' : '#9370DB',
'ORDINAL' : '#9370DB',
}
nlp = spacy.load(_model)
doc = nlp(_text)
html = displacy.render([doc], style = 'ent', options = _options ,jupyter = False)
```

**Reply by ljvmiranda921** [reply]
I am not quite sure how you're constructing the options. Ideally you want something like this:

```
_options = {"colors": _colors_dict, "template": _template}
```

And then you should be able to pass them in the `options` parameter of `displacy.render`. There also seems a problem in your colors dictionary. I don't think `Dark #006400` is valid.
